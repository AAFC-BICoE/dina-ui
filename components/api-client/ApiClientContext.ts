import Kitsu from "kitsu";
import React from "react";
import {
  JsonApiErrorResponse,
  JsonApiResponse,
  Operation,
  OperationsResponse
} from "./jsonapi-types";

/** Api context interface. */
export interface ApiClientContextI {
  /** Client to talk to the back-end API. */
  apiClient: Kitsu;

  /** Function to perform requests against a jsonpatch-compliant JSONAPI server. */
  doOperations: (operations: Operation[]) => Promise<JsonApiResponse[]>;
}

/**
 * React context that passes down a single API client to subscribed components.
 */
export const ApiClientContext = React.createContext<ApiClientContextI>(
  undefined
);

/**
 * Creates the value of the API client context. The app should only need to call this function
 * once to initialize the context.
 */
export function createContextValue(): ApiClientContextI {
  const apiClient = new Kitsu({
    baseURL: "/api",
    pluralize: false,
    resourceCase: "none"
  });

  /**
   * Performs a write operation against a jsonpatch-compliant JSONAPI server.
   */
  async function doOperations(
    operations: Operation[]
  ): Promise<JsonApiResponse[]> {
    // Unwrap the configured axios instance from the Kitsu instance.
    const { axios } = apiClient;

    // Do the operations request.
    const axiosResponse = await axios.patch("operations", operations, {
      headers: {
        Accept: "application/json-patch+json",
        "Content-Type": "application/json-patch+json"
      }
    });

    // Check for errors. At least one error means that the entire request's transaction was
    // cancelled.
    const errorMessage = getErrorMessage(axiosResponse.data);

    // If there is an error message, throw it.
    if (errorMessage) {
      throw new Error(errorMessage);
    }

    // Return the successful jsonpatch response.
    return axiosResponse.data;
  }

  return {
    apiClient,
    doOperations
  };
}

/** Gets the error message as a string from the JSONAPI jsonpatch/operations response. */
function getErrorMessage(
  operationsResponse: OperationsResponse
): string | null {
  // Filter down to just the error responses.
  const errorResponses = operationsResponse.filter(
    ({ status }) => !/2../.test(status.toString())
  ) as JsonApiErrorResponse[];

  // Map the error responses to JsonApiErrors.
  const jsonApiErrors = errorResponses.map(response => response.errors);

  // Convert the JsonApiErrors to an aggregated error string.
  const message = jsonApiErrors
    .map(errors =>
      errors.map(({ title, detail }) => `${title}: ${detail}`).join("\n")
    )
    .join("\n");

  // Return the error message if there is one, or null otherwise.
  return message || null;
}
