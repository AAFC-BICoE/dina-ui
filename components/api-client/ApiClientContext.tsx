import Kitsu from "kitsu";
import React from "react";

/** Api context interface. */
export interface ApiClientContextI {
  apiClient: Kitsu;
}

/**
 * React context that passes down a single API client to subscribed components.
 */
export const ApiClientContext = React.createContext<ApiClientContextI>(
  undefined
);
