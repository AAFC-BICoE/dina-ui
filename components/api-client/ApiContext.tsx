import Kitsu from "kitsu";
import React from "react";

/** Api context interface. */
export interface ApiContextI {
  apiClient: Kitsu;
}

/**
 * React context that passes down a single API client to subscribed components.
 */
export const ApiContext = React.createContext<ApiContextI>(undefined);
