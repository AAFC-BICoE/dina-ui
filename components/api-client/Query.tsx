import { KitsuResponseData } from "kitsu";
import React from "react";
import { JsonApiQuerySpec, QueryState, useQuery } from "./useQuery";

/** Query component props. */
export interface QueryProps<TData extends KitsuResponseData, TMeta> {
  query: JsonApiQuerySpec;
  children: QueryChildren<TData, TMeta>;
}

/**
 * Query component's "children" prop type.
 *
 * This component uses the render props pattern to pass data to children components.
 * See: https://reactjs.org/docs/render-props.html
 */
type QueryChildren<TData extends KitsuResponseData, TMeta> = (
  state: QueryState<TData, TMeta>
) => React.ReactElement;

/** Back-end connected Query component. */
export function Query<TData extends KitsuResponseData, TMeta = undefined>({
  children,
  query
}: QueryProps<TData, TMeta>) {
  const queryState = useQuery<TData, TMeta>(query);
  return children(queryState);
}
