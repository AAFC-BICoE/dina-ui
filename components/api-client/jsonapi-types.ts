// JSONAPI types.

/** A sucessful response from a JSONAPI backend. */
export interface JsonApiResponse {
  data: JsonApiResource;
  status: 201;
}

/** A single entity in JSONAPI format. */
export interface JsonApiResource {
  id: number | string;
  type: string;
  attributes?: any;
  relationships?: any;
}

/**
 * JSONAPI error response.
 * See https://jsonapi.org/format/#errors
 */
export interface JsonApiErrorResponse {
  errors: JsonApiError[];
  status: number;
}

/**
 * JSONAPI error object.
 * See https://jsonapi.org/format/#error-objects
 */
export interface JsonApiError {
  id?: string;
  links?: any;
  status?: string;
  code?: string;
  title?: string;
  detail?: string;
  source?: any;
  meta?: any;
}

/** HTTP method used to specify a jsonpatch operation type. */
export type HttpMethod = "POST" | "PATCH" | "DELETE";

/**
 * A jsonpatch operation.
 *
 * See: https://github.com/json-api/json-api/blob/9c7a03dbc37f80f6ca81b16d444c960e96dd7a57/extensions/jsonpatch/index.md#-creating-resources
 */
export interface Operation {
  op: HttpMethod;
  path: string;
  value: JsonApiResource;
}

/**
 * A response from a JSONAPI jsonpatch backend.
 *
 * See: https://github.com/json-api/json-api/blob/9c7a03dbc37f80f6ca81b16d444c960e96dd7a57/extensions/jsonpatch/index.md#-responses
 */
export type OperationsResponse = Array<JsonApiResponse | JsonApiErrorResponse>;
