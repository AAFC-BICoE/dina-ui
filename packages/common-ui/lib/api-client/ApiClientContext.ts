import Kitsu, { KitsuResource, PersistedResource } from "kitsu";
import { deserialise } from "kitsu-core";
import React from "react";
import { serialize } from "../util/serialize";
import { ClientSideJoiner, ClientSideJoinSpec } from "./client-side-join";
import {
  FailedOperation,
  Operation,
  OperationsResponse,
  SuccessfulOperation
} from "./operations-types";

export interface BulkGetOptions {
  apiBaseUrl?: string;
  joinSpecs?: ClientSideJoinSpec[];

  /** Return null for missing resource instead of throwing an Error. */
  returnNullForMissingResource?: boolean;
}

export interface DoOperationsOptions {
  apiBaseUrl?: string;

  /** Return null for missing resource instead of throwing an Error. */
  returnNullForMissingResource?: boolean;
}

/** Api context interface. */
export interface ApiClientContextI {
  /** Client to talk to the back-end API. */
  apiClient: Kitsu;

  /** Function to perform requests against a jsonpatch-compliant JSONAPI server. */
  doOperations: (
    operations: Operation[],
    options?: DoOperationsOptions
  ) => Promise<SuccessfulOperation[]>;

  /** Creates or updates one or multiple resources. */
  save: (
    saveArgs: SaveArgs[],
    options?: DoOperationsOptions
  ) => Promise<Array<PersistedResource<KitsuResource>>>;

  /** Bulk GET operations: Run many find-by-id queries in a single HTTP request. */
  bulkGet: <T extends KitsuResource>(
    paths: string[],
    options?: BulkGetOptions
  ) => Promise<Array<PersistedResource<T>>>;
}

/** Config for creating an API client context value. */
export interface ApiClientContextConfig {
  /** Back-end API base URL. */
  baseURL?: string;

  /**
   * Temp ID iterator.
   * This is not persisted on the back-end as the actual database ID.
   * The generated ID should be in the back-end's expected format (e.g. number or UUID).
   */
  getTempIdGenerator?: () => () => string;

  headers?: Record<string, string>;
}

/** save function args. */
export interface SaveArgs<T extends KitsuResource = KitsuResource> {
  resource: T;
  type: string;
}

/**
 * React context that passes down a single API client to subscribed components.
 */
export const ApiClientContext = React.createContext<ApiClientContextI>(
  // Default value is undefined. This won't matter as long as the hook is called inside the context provider.
  undefined as any
);

/**
 * Creates the value of the API client context. The app should only need to call this function
 * once to initialize the context.
 */
export function createContextValue({
  baseURL = "/api",
  getTempIdGenerator = () => {
    let idIterator = -100;
    return () => String(idIterator--);
  },
  headers = {}
}: ApiClientContextConfig = {}): ApiClientContextI {
  const apiClient = new Kitsu({
    baseURL,
    headers: { "Crnk-Compact": "true", ...headers },
    pluralize: false,
    resourceCase: "none"
  });

  /**
   * Performs a write operation against a jsonpatch-compliant JSONAPI server.
   */
  async function doOperations(
    operations: Operation[],
    { apiBaseUrl = "", returnNullForMissingResource }: DoOperationsOptions = {}
  ): Promise<SuccessfulOperation[]> {
    // Unwrap the configured axios instance from the Kitsu instance.
    const { axios } = apiClient;

    // Do the operations request.
    const axiosResponse = await axios.patch(
      `${apiBaseUrl}/operations`,
      operations,
      {
        headers: {
          Accept: "application/json-patch+json",
          "Content-Type": "application/json-patch+json",
          "Crnk-Compact": "true"
        }
      }
    );

    const responses: OperationsResponse = axiosResponse.data;

    // Optionally return null instead of throwing an error for missing resources:
    if (returnNullForMissingResource) {
      for (const i in responses) {
        // 404 Not Found or 410 Gone
        if ([404, 410].includes(responses[i].status)) {
          responses[i] = {
            data: null as any,
            status: 404
          };
        }
      }
    }

    // Check for errors. At least one error means that the entire request's transaction was
    // cancelled.
    const errorMessage = getErrorMessage(responses);

    // If there is an error message, throw it.
    if (errorMessage) {
      throw new Error(errorMessage);
    }

    // Return the successful jsonpatch response.
    return responses as SuccessfulOperation[];
  }

  /**
   * Creates or updates one or multiple resources.
   */
  async function save(
    saveArgs: SaveArgs[],
    options?: DoOperationsOptions
  ): Promise<Array<PersistedResource<KitsuResource>>> {
    // Serialize the resources to JSONAPI format.
    const serializePromises = saveArgs.map(saveArg => serialize(saveArg));
    const serialized = await Promise.all(serializePromises);

    // Temp ID iterator. This is not persisted on the back-end as the actual database ID.
    const generateId = getTempIdGenerator();

    // Create the jsonpatch oeprations objects.
    const operations = serialized.map<Operation>(jsonapiResource => ({
      op: jsonapiResource.id ? "PATCH" : "POST",
      path: jsonapiResource.id
        ? `${jsonapiResource.type}/${jsonapiResource.id}`
        : jsonapiResource.type,
      value: {
        ...jsonapiResource,
        id: String(jsonapiResource.id || generateId())
      }
    }));

    // Do the operations request.
    const responses = await doOperations(operations, options);

    // Deserialize the responses to Kitsu format.
    const deserializePromises = responses.map(response =>
      deserialise(response)
    );
    const deserialized = await Promise.all(deserializePromises);
    const kitsuResources = deserialized.map(({ data }) => data);

    return kitsuResources;
  }

  /** Bulk GET operations: Run many find-by-id queries in a single HTTP request. */
  async function bulkGet<T extends KitsuResource>(
    paths: string[],
    {
      apiBaseUrl = "",
      joinSpecs = [],
      returnNullForMissingResource
    }: BulkGetOptions = {}
  ) {
    const getOperations = paths.map<Operation>(path => ({
      op: "GET",
      path
    }));

    const responses = await doOperations(getOperations, {
      apiBaseUrl,
      returnNullForMissingResource
    });

    const resources: Array<PersistedResource<T>> = (
      await Promise.all(responses.map(deserialise))
    ).map(res => res.data);

    for (const joinSpec of joinSpecs) {
      await new ClientSideJoiner(bulkGet, resources, joinSpec).join();
    }

    return resources;
  }

  return {
    apiClient,
    bulkGet,
    doOperations,
    save
  };
}

/** Gets the error message as a string from the JSONAPI jsonpatch/operations response. */
function getErrorMessage(
  operationsResponse: OperationsResponse
): string | null {
  // Filter down to just the error responses.
  const errorResponses = operationsResponse.filter(
    ({ status }) => !/2../.test(status.toString())
  ) as FailedOperation[];

  // Map the error responses to JsonApiErrors.
  const jsonApiErrors = errorResponses
    // Ignore any error responses without an 'errors' field.
    .filter(response => response.errors)
    .map(response => response.errors);

  // Convert the JsonApiErrors to an aggregated error string.
  const message = jsonApiErrors
    .map(errors =>
      errors.map(({ title, detail }) => `${title}: ${detail}`).join("\n")
    )
    .join("\n");

  // Return the error message if there is one, or null otherwise.
  return message || null;
}
