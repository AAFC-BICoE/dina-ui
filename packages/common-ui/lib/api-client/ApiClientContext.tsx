import { AxiosError } from "axios";
import { cacheAdapterEnhancer } from "axios-extensions";
import { FormikErrors } from "formik";
import Kitsu, {
  GetParams,
  InputResource,
  KitsuResource,
  KitsuResourceLink,
  PersistedResource
} from "kitsu";
import { deserialise, error as kitsuError, query } from "kitsu-core";
import { compact, flatMap, fromPairs, isEmpty, keys, omit } from "lodash";
import LRUCache from "lru-cache";
import React, { PropsWithChildren, useContext, useMemo } from "react";
import { v4 as uuidv4 } from "uuid";
import { OperationsResponse } from "..";
import { serialize } from "../util/serialize";
import { ClientSideJoiner, ClientSideJoinSpec } from "./client-side-join";
import {
  FailedOperation,
  Operation,
  SuccessfulOperation
} from "./operations-types";
import DataLoader from "dataloader";

export interface BulkGetOptions<TReturnNull extends boolean = false> {
  apiBaseUrl?: string;
  joinSpecs?: ClientSideJoinSpec[];

  /** Return null for missing resource instead of throwing an Error. */
  returnNullForMissingResource?: TReturnNull;
}

export type BulkGetOperation =
  | SuccessfulOperation
  | FailedOperation
  // For replacing missing/deleted data with null locally:
  | { data: null; status: 404 };

export interface DoOperationsOptions {
  apiBaseUrl?: string;

  /** Return null for missing resource instead of throwing an Error. */
  returnNullForMissingResource?: boolean;

  overridePatchOperation?: boolean;
}

/** Api client interface. */
export interface ApiClientI {
  /** Client to talk to the back-end API. */
  apiClient: Kitsu;

  /** Function to perform requests against a jsonpatch-compliant JSONAPI server. */
  doOperations: (
    operations: Operation[],
    options?: DoOperationsOptions
  ) => Promise<SuccessfulOperation[]>;

  /** Creates or updates one or multiple resources. */
  save: <TData extends KitsuResource = KitsuResource>(
    args: (SaveArgs | DeleteArgs)[],
    options?: DoOperationsOptions
  ) => Promise<PersistedResource<TData>[]>;

  /** Bulk GET operations: Run many find-by-id queries in a single HTTP request. */
  bulkGet: <T extends KitsuResource, TReturnNull extends boolean = false>(
    paths: readonly string[],
    options?: BulkGetOptions<TReturnNull>
  ) => Promise<
    (TReturnNull extends true
      ? PersistedResource<T> | null
      : PersistedResource<T>)[]
  >;
}

/** Config for creating an API client. */
export interface ApiClientConfig {
  /** Back-end API base URL. */
  baseURL?: string;
  /** New Id generator. */
  newId?: () => string;
}

/** save function args. */
export interface SaveArgs<T extends KitsuResource = KitsuResource> {
  resource: T | InputResource<T>;
  type: string;
}

export interface DeleteArgs {
  delete: KitsuResourceLink | PersistedResource<any>;
}

/**
 * React Context that passes down the API client to Components.
 */
export const ApiClientContext = React.createContext<ApiClientI>(
  // Default value is undefined. This won't matter as long as the hook is called inside the context provider.
  undefined as any
);

export const ApiClientProvider = ApiClientContext.Provider;

export function ApiClientImplProvider({
  children,
  ...cfg
}: PropsWithChildren<ApiClientConfig> = {}) {
  const apiContext = useMemo(() => new ApiClientImpl(cfg), []);
  return <ApiClientProvider value={apiContext}>{children}</ApiClientProvider>;
}

/** Hook to get the Api Client Context */
export function useApiClient() {
  const ctx = useContext(ApiClientContext);
  if (!ctx) {
    throw new Error("No ApiClientContext available.");
  }
  return ctx;
}

export class ApiClientImpl implements ApiClientI {
  public apiClient: Kitsu;

  constructor(private cfg: ApiClientConfig = {}) {
    this.apiClient = new CustomDinaKitsu({
      baseURL: cfg.baseURL ?? "/api",
      headers: { "Crnk-Compact": "true" },
      pluralize: false,
      resourceCase: "none"
    });

    this.apiClient.axios?.interceptors?.response.use(
      (successResponse) => successResponse,
      makeAxiosErrorMoreReadable
    );

    if (this.apiClient.axios?.defaults?.adapter) {
      const ONE_SECOND = 1000;
      this.apiClient.axios.defaults.adapter = cacheAdapterEnhancer(
        this.apiClient.axios.defaults.adapter,
        {
          // Invalidate the cache after one second.
          // All this does is batch requests if a set of react components all try to make the same request at once.
          // e.g. a page with a lot of the same dropdown select component, or a set of group label components fetching the label for the same group.
          defaultCache: new LRUCache({ max: 100, maxAge: ONE_SECOND })
        }
      );
    }

    // Bind the methods so context consumers can use object destructuring.
    this.doOperations = this.doOperations.bind(this);
    this.save = this.save.bind(this);
    this.bulkGet = this.bulkGet.bind(this);
  }

  /**
   * Performs a write operation against a jsonpatch-compliant JSONAPI server.
   */
  public async doOperations(
    operations: Operation[],
    { apiBaseUrl = "", returnNullForMissingResource }: DoOperationsOptions = {}
  ): Promise<SuccessfulOperation[]> {
    // Unwrap the configured axios instance from the Kitsu instance.
    const { axios } = this.apiClient;

    // Do the operations request.
    const axiosResponse = await axios.patch(
      `${apiBaseUrl}/operations`,
      operations,
      {
        headers: {
          Accept: "application/json-patch+json",
          "Content-Type": "application/json-patch+json",
          "Crnk-Compact": "true"
        }
      }
    );

    const responses: OperationsResponse | BulkGetOperation[] =
      axiosResponse.data;

    // Optionally return null instead of throwing an error for missing resources:
    if (returnNullForMissingResource) {
      for (const i in responses) {
        // 404 Not Found or 410 Gone
        if ([404, 410].includes(responses[i].status)) {
          responses[i] = {
            data: null,
            status: 404
          };
        }
      }
    }

    // Check for errors. At least one error means that the entire request's transaction was
    // cancelled.
    const { errorMessage, fieldErrors, individualErrors } =
      getErrorMessages(responses);

    // If there is an error message, throw it.
    if (errorMessage || !isEmpty(fieldErrors)) {
      throw new DoOperationsError(
        errorMessage ?? "",
        fieldErrors,
        individualErrors
      );
    }

    // Return the successful jsonpatch response.
    return responses as SuccessfulOperation[];
  }

  /**
   * Creates or updates one or multiple resources.
   */
  public async save<TData extends KitsuResource = KitsuResource>(
    args: (SaveArgs | DeleteArgs)[],
    options?: DoOperationsOptions
  ): Promise<PersistedResource<TData>[]> {
    const deleteArgs = args.filter(
      (arg) => (arg as any).delete
    ) as DeleteArgs[];
    const saveArgs = args.filter((arg) => !(arg as any).delete) as SaveArgs[];

    // Serialize the resources to JSONAPI format.
    const serializePromises = saveArgs.map((saveArg) => serialize(saveArg));
    const serialized = await Promise.all(serializePromises);

    // Create the jsonpatch operations objects.
    const saveOperations = serialized.map<Operation>((jsonapiResource) => ({
      op: options?.overridePatchOperation
        ? "POST"
        : jsonapiResource.id
        ? "PATCH"
        : "POST",
      path: options?.overridePatchOperation
        ? jsonapiResource.type
        : jsonapiResource.id
        ? `${jsonapiResource.type}/${jsonapiResource.id}`
        : jsonapiResource.type,
      value: {
        ...jsonapiResource,
        id: String(jsonapiResource.id || this.cfg.newId?.() || uuidv4())
      }
    }));

    const deleteOperations = deleteArgs.map<Operation>((deleteArg) => ({
      op: "DELETE",
      path: `${deleteArg.delete.type}/${deleteArg.delete.id}`
    }));

    const operations = [...saveOperations, ...deleteOperations];

    // Do the operations request.
    const responses = await this.doOperations(operations, options);

    // Deserialize the responses to Kitsu format.
    const deserializePromises = responses.map((response) =>
      deserialise(response)
    );
    const deserialized = await Promise.all(deserializePromises);
    const kitsuResources = deserialized.map(({ data }) => data);

    return kitsuResources;
  }

  /** Bulk GET operations: Run many find-by-id queries in a single HTTP request. */
  public async bulkGet<
    T extends KitsuResource,
    TReturnNull extends boolean = false
  >(
    paths: string[],
    {
      apiBaseUrl = "",
      joinSpecs = [],
      returnNullForMissingResource
    }: BulkGetOptions<TReturnNull> = {}
  ) {
    // Don't do an empty operations request:
    if (!paths.length) {
      return [];
    }

    // Use DataLoader to avoid requesting the same ID multiple times,
    // which crnk-operations throws an error for:
    const batchLoader = new DataLoader<string, SuccessfulOperation>(
      async (uniquePaths) => {
        const getOperations = uniquePaths.map<Operation>((path) => ({
          op: "GET",
          path
        }));

        return await this.doOperations(getOperations, {
          apiBaseUrl,
          returnNullForMissingResource
        });
      }
    );

    const responses = await batchLoader.loadMany(paths);

    const resources: (TReturnNull extends true
      ? PersistedResource<T> | null
      : PersistedResource<T>)[] = (
      await Promise.all(responses.map(deserialise))
    ).map((res) => res.data);

    for (const joinSpec of joinSpecs) {
      await new ClientSideJoiner(this.bulkGet, resources, joinSpec).join();
    }

    return resources;
  }
}

export interface OperationError {
  index: number | string;
  errorMessage: string | null;
  fieldErrors: FormikErrors<any>;
}

/** Gets the error message as a string from the JSONAPI jsonpatch/operations response. */
export function getErrorMessages(
  operationsResponse: OperationsResponse | BulkGetOperation[]
): {
  errorMessage: string | null;
  fieldErrors: FormikErrors<any>;
  /** The error messages for each indivisual operation with the operation's index. */
  individualErrors: OperationError[];
} {
  // Filter down to just the error responses.
  const errorResponses = operationsResponse
    .map((res, index: number) => ({ index, response: res as FailedOperation }))
    .filter(({ response }) => !/2../.test(response.status.toString()));

  const individualErrors = errorResponses.map(({ response, index }) => {
    // Map the error responses to JsonApiErrors.
    // Ignore any error responses without an 'errors' field.
    const jsonApiErrors = compact(response.errors);

    // Convert the JsonApiErrors to an aggregated error string.
    const errorMessage =
      jsonApiErrors
        // Don't include field-level errors in the form-level error message:
        .filter((error) => !error.source?.pointer)
        .map(
          // The error message is the title + detail, but remove one if the other is missing
          ({ title, detail }) =>
            [title, detail].filter((s) => s?.trim()).join(": ")
        )
        .join("\n") || null;

    const fieldErrors = fromPairs(
      jsonApiErrors
        // Only include field-level errors in the fieldErrors:
        .filter((error) => error.source?.pointer && error.detail)
        .map((error) => [
          error.source?.pointer?.toString?.() ?? "",
          error.detail ?? ""
        ])
    );

    return { index, errorMessage, fieldErrors };
  });

  const overallErrorMessage =
    compact(individualErrors.map((it) => it.errorMessage)).join("\n") || null;
  const overallFieldErrors = individualErrors.reduce(
    (total, curr) => ({ ...total, ...curr.fieldErrors }),
    {}
  );

  // Return the error message if there is one, or null otherwise.
  return {
    errorMessage: overallErrorMessage,
    fieldErrors: overallFieldErrors,
    individualErrors
  };
}

/** Error class thrown by doOperations function. */
export class DoOperationsError extends Error {
  constructor(
    public message: string,
    public fieldErrors: FormikErrors<any> = {},
    public individualErrors: OperationError[] = [
      { errorMessage: message, fieldErrors, index: 0 }
    ]
  ) {
    super(message);
  }
}

/** Show more details in the Axios errors. */
export function makeAxiosErrorMoreReadable(error: AxiosError) {
  if (error.isAxiosError) {
    let errorMessage = `${error.config.url}: ${error.response?.statusText}`;

    // Special case: Make 502 "bad gateway" messages more user-friendly:
    if (error.response?.status === 502) {
      errorMessage = `Service unavailable:\n${errorMessage}`;
    }

    // Handle errors coming from Spring Boot:
    if (error.response?.data?.errors) {
      const jsonApiErrorResponse = {
        status: error.response.status,
        errors: error.response.data.errors
      };
      errorMessage +=
        "\n" + getErrorMessages([jsonApiErrorResponse]).errorMessage;
    }

    throw new Error(errorMessage);
  }
  throw error;
}

export class CustomDinaKitsu extends Kitsu {
  /**
   * The default Kitsu 'get' method omits the last part of URLs with multiple slashes.
   * e.g. "seqdb-api/index-set/1/ngsindexes" becomes "seqdb-api/index-set/1".
   * Override the 'get' method so it works with our long URLs:
   */
  async get(path: string, params: GetParams = {}) {
    const paramsNet = omit(params, "header");
    try {
      const { data } = await this.axios.get(path, {
        headers: { ...this.headers, ...params.header },
        params: paramsNet,
        paramsSerializer: (p) => query(p)
      });

      const deserialized = await deserialise(data);

      // Omit relationships where: { data: null } because they do not deserialize properly:
      const relationships = deserialized?.data?.relationships;
      for (const key of keys(relationships)) {
        if (relationships?.[key]?.data === null) {
          delete relationships[key];
        }
      }

      return deserialized;
    } catch (E) {
      throw kitsuError(E);
    }
  }
}
