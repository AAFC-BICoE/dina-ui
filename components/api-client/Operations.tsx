import React from "react";
import { ApiClientContext, ApiClientContextI } from "./ApiClientContext";

/** HTTP method used to specify a jsonpatch operation type. */
type HttpMethod = "POST" | "PATCH" | "DELETE";

/**
 * A jsonpatch operation.
 *
 * See: https://github.com/json-api/json-api/blob/9c7a03dbc37f80f6ca81b16d444c960e96dd7a57/extensions/jsonpatch/index.md#-creating-resources
 */
export interface Operation {
  op: HttpMethod;
  path: string;
  value: JsonApiResource;
}

/** A single entity in JSONAPI format. */
interface JsonApiResource {
  id: number | string;
  type: string;
  attributes?: any;
  relationships?: any;
}

/**
 * "Render props" values that are passed down to children components.
 */
interface OperationsRenderProps {
  doOperations: (operations: Operation[]) => void;
}

/**
 * Operations component's "children" prop type.
 */
type OperationsChildren = (
  renderProps: OperationsRenderProps
) => React.ReactNode;

/** Operations component props. */
interface OperationsProps {
  children: OperationsChildren;
}

/**
 * Provides an interface for performing write operations against the backend via render props.
 * See: https://reactjs.org/docs/render-props.html
 *
 * The backend must support the jsonpatch extension to JSONAPI.
 * See: https://github.com/json-api/json-api/blob/9c7a03dbc37f80f6ca81b16d444c960e96dd7a57/extensions/jsonpatch/index.md
 */
export class Operations extends React.Component<OperationsProps> {
  static contextType = ApiClientContext;

  /**
   * Performs write operations against the backend.
   *
   * @param operations the jsonpatch operations to perform
   */
  private async doOperations(operations: Operation[]) {
    // Unwrap the configured axios instance from the Kitsu instance.
    const {
      apiClient: { axios }
    } = this.context as ApiClientContextI;

    // Do the operations request.
    await axios.patch("operations", operations, {
      headers: {
        "Content-Type": "application/json-patch+json",
        Accept: "application/json-patch+json"
      }
    });
  }

  render() {
    return this.props.children({
      doOperations: ops => this.doOperations(ops)
    });
  }
}
