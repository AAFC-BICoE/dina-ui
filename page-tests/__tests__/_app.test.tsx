import { mount } from "enzyme";
import Kitsu from "kitsu";
import { FunctionComponent } from "react";
import { ApiClientContext } from "../../components/api-client/ApiClientContext";
import SeqdbUiApp from "../../pages/_app";

const mockPush = jest.fn();

const mockRouter = {
  asPath: "/example-path?a=b",
  push: mockPush
};

describe("SeqdbUiApp", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("Renders the App wrapper.", async () => {
    const TestComponent: FunctionComponent = () => <div />;

    const wrapper = mount(
      <SeqdbUiApp
        router={mockRouter as any}
        pageProps={{ exampleProp: "exampleValue" }}
        Component={TestComponent}
      />
    );

    // The first browser render should be empty.
    expect(wrapper.html()).toEqual(null);

    expect(mockPush).lastCalledWith("/example-path?a=b");
    // Normally the router would update the app wrapper, but the mock doesn't, so we force a
    // re-render in the test.
    wrapper.instance().forceUpdate();
    wrapper.update();

    const innerComponent = wrapper.find(TestComponent);
    expect(innerComponent.prop("exampleProp")).toEqual("exampleValue");
  });

  it("Provides the API context to child components.", done => {
    function pageComponent() {
      return (
        <ApiClientContext.Consumer>
          {context => {
            expect(context.apiClient instanceof Kitsu).toBeTruthy();
            expect(typeof context.doOperations).toEqual("function");
            done();
            return null;
          }}
        </ApiClientContext.Consumer>
      );
    }

    const wrapper = mount(
      <SeqdbUiApp
        router={mockRouter as any}
        pageProps={{}}
        Component={pageComponent}
      />
    );

    expect(mockPush).lastCalledWith("/example-path?a=b");
    // Normally the router would update the app wrapper, but the mock doesn't, so we force a
    // re-render in the test.
    wrapper.instance().forceUpdate();
    wrapper.update();
  });
});
