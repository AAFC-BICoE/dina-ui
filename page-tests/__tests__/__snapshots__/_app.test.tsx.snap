// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SeqdbUiApp Renders the App wrapper.: SeqdbUiApp shallow render. 1`] = `
<Context.Provider
  value={
    Object {
      "apiClient": Kitsu {
        "axios": [Function],
        "camel": [Function],
        "create": [Function],
        "fetch": [Function],
        "headers": Object {
          "Accept": "application/vnd.api+json",
          "Content-Type": "application/vnd.api+json",
        },
        "plural": [Function],
        "remove": [Function],
        "resCase": [Function],
        "update": [Function],
      },
      "doOperations": [Function],
    }
  }
>
  <Container>
    <TestComponent
      exampleProp="exampleValue"
    />
  </Container>
</Context.Provider>
`;
